crates/rwasm/machine/src/alu/mul64/mod.rs

```diff
@@ -0,0 +1,414 @@
+
+        blu.add_byte_lookup_events(vec![
+            ByteLookupEvent {
+                opcode: ByteOpcode::MSB,
+                a1: b_msb as u16,
+                a2: 0,
+                b: b_word[WORD_SIZE - 1],
+                c: 0,
+            },
+            ByteLookupEvent {
+                opcode: ByteOpcode::MSB,
+                a1: c_msb as u16,
+                a2: 0,
+                b: c_word[WORD_SIZE - 1],
+                c: 0,
+            },
+        ]);
+
+        let mut product = [0u32; LONG_WORD_SIZE];
+        for i in 0..b.len() {
+            for j in 0..c.len() {
+                if i + j < LONG_WORD_SIZE {
+                    product[i + j] += (b[i] as u32) * (c[j] as u32);
+                }
+            }
+        }
+
+        let base = (1 << BYTE_SIZE) as u32;
+        let mut carry = [0u32; LONG_WORD_SIZE];
+        for i in 0..LONG_WORD_SIZE {
+            carry[i] = product[i] / base;
+            product[i] %= base;
+            if i + 1 < LONG_WORD_SIZE {
+                product[i + 1] += carry[i];
+            }
+            cols.carry[i] = F::from_canonical_u32(carry[i]);
+        }
+
+        cols.product = product.map(F::from_canonical_u32);
+        cols.a_lo = Word(a_lo_word.map(F::from_canonical_u8));
+        cols.a_hi = Word(a_hi_word.map(F::from_canonical_u8));
+        cols.b = Word(b_word.map(F::from_canonical_u8));
+        cols.c = Word(c_word.map(F::from_canonical_u8));
+        cols.is_real = F::one();
+        cols.is_mul64 = F::one();
+
+        blu.add_u16_range_checks(&carry.map(|x| x as u16));
+        blu.add_u8_range_checks(&product.map(|x| x as u8));
+    }
+}
+
+impl<AB> Air<AB> for Mul64Chip
+where
+    AB: SP1CoreAirBuilder,
+{
+    fn eval(&self, builder: &mut AB) {
+        let main = builder.main();
+        let local = main.row_slice(0);
+        let local: &Mul64Cols<AB::Var> = (*local).borrow();
+        let base = AB::F::from_canonical_u32(1 << 8);
+        let zero: AB::Expr = AB::F::zero().into();
+        let byte_mask = AB::F::from_canonical_u8(BYTE_MASK);
+
+        builder.assert_bool(local.is_real);
+        builder.assert_bool(local.is_mul64);
+        builder.when(local.is_real).assert_one(local.is_mul64);
+
+        // MSB checks
+        builder.send_byte(
+            ByteOpcode::MSB.as_field::<AB::F>(),
+            local.b_msb,
+            local.b[WORD_SIZE - 1],
+            zero.clone(),
+            local.is_real,
+        );
+        builder.send_byte(
+            ByteOpcode::MSB.as_field::<AB::F>(),
+            local.c_msb,
+            local.c[WORD_SIZE - 1],
+            zero.clone(),
+            local.is_real,
+        );
+
+        // Sign extension calculation
+        builder.assert_eq(local.b_sign_extend, local.b_msb);
+        builder.assert_eq(local.c_sign_extend, local.c_msb);
+        builder.assert_bool(local.b_sign_extend);
+        builder.assert_bool(local.c_sign_extend);
+
+        // Sign extend b and c
+        let (b, c) = {
```

Comment:
the result can be constrained by the Mul chip. Not need to redo the multiplication.